src/optimize.rs

```diff
@@ -0,0 +1,1049 @@
+                    borrowed(&mut new_sections, context.full_wasm, data, SectionId::Data)
+                }
+
+                Payload::ModuleSection { .. }
+                | Payload::InstanceSection(_)
+                | Payload::CoreTypeSection(_)
+                | Payload::ComponentSection { .. }
+                | Payload::ComponentInstanceSection(_)
+                | Payload::ComponentAliasSection(_)
+                | Payload::ComponentTypeSection(_)
+                | Payload::ComponentCanonicalSection(_)
+                | Payload::ComponentStartSection { .. }
+                | Payload::ComponentImportSection(_)
+                | Payload::ComponentExportSection(_) => {
+                    unreachable!("component model not supported yet; disabled in validator")
+                }
+
+                Payload::UnknownSection {
+                    id,
+                    contents: _,
+                    range,
+                } => bail!("unknown section with id {id} at range {range:?}"),
+
+                Payload::End(_) => break,
+            }
+
+            wasm = &wasm[consumed..];
+        }
+
+        let mut new_code_section = Some(self.optimize_func_bodies(&mut context)?);
+
+        log::trace!("Building final optimized module");
+        let mut module = wasm_encoder::Module::new();
+        for section in &new_sections {
+            use wasm_encoder::Section;
+
+            if context
+                .new_global_section
+                .as_ref()
+                .map_or(false, |s| s.id() < section.id())
+            {
+                let s = context.new_global_section.take().unwrap();
+                log::trace!("Appending section id: {}", s.id());
+                module.section(&s);
+            }
+
+            if context
+                .new_export_section
+                .as_ref()
+                .map_or(false, |s| s.id() < section.id())
+            {
+                let s = context.new_export_section.take().unwrap();
+                log::trace!("Appending section id: {}", s.id());
+                module.section(&s);
+            }
+
+            if new_code_section.as_ref().map_or(false, |s| {
+                s.id() < section.id() && section.id() != SectionId::DataCount as u8
+            }) {
+                let s = new_code_section.take().unwrap();
+                log::trace!("Appending section id: {}", s.id());
+                module.section(&s);
+            }
+
+            log::trace!("Appending section id: {}", section.id());
+            module.section(&*section);
+        }
+        if let Some(s) = new_code_section {
+            module.section(&s);
+        }
+        Ok(module.finish())
+    }
+
+    fn optimize_func_bodies(
+        &self,
+        context: &mut OptimizeContext,
+    ) -> Result<wasm_encoder::CodeSection> {
+        let mut call_site_index = 0;
+        for body in context.func_bodies.iter() {
+            context.call_site_offsets.push(call_site_index);
```

Comment:
Maybe call this `first_call_site_offset_for_func` or something, to make it a bit more clear?